sql: panic on UPDATE: interface conversion

See #7861, #7863.

SQL:

```
CREATE DATABASE d;
SET DATABASE = d;
CREATE TABLE t();
UPDATE t SET ident.* = DEFAULT;
```

Panic:

```
panic: interface conversion: parser.IndirectionElem is parser.StarIndirection, not parser.NameIndirection

goroutine 371 [running]:
github.com/cockroachdb/cockroach/sql/parser.(*QualifiedName).Column(0xc820178b00, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/expr.go:770 +0x1bb
github.com/cockroachdb/cockroach/sql.(*planner).processColumns(0xc8206cc090, 0xc820cc5a00, 0xc82073fc40, 0x3, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/insert.go:341 +0x281
github.com/cockroachdb/cockroach/sql.(*planner).Update(0xc8206cc090, 0xc8207ca5a0, 0x0, 0x0, 0x0, 0x7fd043cbc501, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/update.go:151 +0x446
github.com/cockroachdb/cockroach/sql.(*planner).newPlan(0xc8206cc090, 0x7fd043cc1be0, 0xc8207ca5a0, 0x0, 0x0, 0x0, 0x19a29801, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/plan.go:306 +0x717
github.com/cockroachdb/cockroach/sql.(*planner).makePlan(0xc8206cc090, 0x7fd043cc1be0, 0xc8207ca5a0, 0x6aa501, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/plan.go:212 +0x83
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc820208000, 0x7fd043cc1be0, 0xc8207ca5a0, 0xc8206cc090, 0xc8207ca501, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:1047 +0x93
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtInOpenTxn(0xc820208000, 0x7fd043cc1be0, 0xc8207ca5a0, 0xc8206cc090, 0x101, 0xc8206cc028, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:945 +0x439
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtsInCurrentTxn(0xc820208000, 0xc8206e71f0, 0x1, 0x1, 0xc8206cc090, 0xc8206cc028, 0x2220101, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:693 +0x5da
github.com/cockroachdb/cockroach/sql.runTxnAttempt(0xc820208000, 0xc8206cc090, 0x1, 0xc8206cc028, 0xc82071e420, 0xc8206e71f0, 0x1, 0x1, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:621 +0x18d
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest.func2(0xc8207aedc0, 0xc82071e420, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:513 +0x470
github.com/cockroachdb/cockroach/internal/client.(*Txn).Exec(0xc8207aedc0, 0xc820330101, 0x146250324c035440, 0x0, 0xc820333980, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/internal/client/txn.go:548 +0x2eb
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc820208000, 0x7fd043d3c830, 0xc820756140, 0xc8206cc000, 0xc820862137, 0x9e, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:518 +0xba7
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc820208000, 0x7fd043d3c830, 0xc820756140, 0xc8206cc000, 0xc820862137, 0x9e, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:384 +0x146
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc8201a1500, 0x7fd043d3c830, 0xc820756140, 0xc820862137, 0x9e, 0x0, 0x0, 0x0, 0x0, 0x7fd047c76001, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:640 +0x98
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc8201a1500, 0x7fd043d3c830, 0xc820756140, 0xc8201a1528, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:320 +0xe8
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc8201a1500, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:275 +0x100c
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc820186720, 0x7fd043cfc100, 0xc8200f6420, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:231 +0x9d9
github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1(0x7fd043cfc100, 0xc8200f6420)
	/go/src/github.com/cockroachdb/cockroach/server/server.go:370 +0x42
github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith.func1(0xc8205a2050, 0x7fd043cfc100, 0xc8200f6420, 0xc82039c000)
	/go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:131 +0x62
created by github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith
	/go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:133 +0x333
```

Found with random syntax generator.

--- a/sql/insert.go
+++ b/sql/insert.go
@@ -333,17 +333,23 @@ func (p *planner) processColumns(tableDesc *sqlbase.TableDescriptor,
 	cols := make([]sqlbase.ColumnDescriptor, len(node))
 	colIDSet := make(map[sqlbase.ColumnID]struct{}, len(node))
 	for i, n := range node {
-		// TODO(pmattis): If the name is qualified, verify the table name matches
-		// tableDesc.Name.
-		if err := n.NormalizeColumnName(); err != nil {
-			return nil, err
-		}
-		col, err := tableDesc.FindActiveColumnByName(n.Column())
+		// Qualified names in the column list for INSERT and for the LHS
+		// in UPDATE ... SET statements work differently than qualified
+		// column names elsewhere:
+		// - they always start with the column name,
+		// - notation X.Y means access field Y when column X has composite type,
+		// - notation X.* is never allowed.
+		// So NormalizeColumnName() cannot be used here.
+		col, err := tableDesc.FindActiveColumnByName(string(n.Base))
 		if err != nil {
 			return nil, err
 		}
+		if len(n.Indirect) != 0 {
+			// We do not support anything but simple column names yet.
+			return nil, fmt.Errorf("column name \"%s\" cannot be assigned to", n)
+		}
 		if _, ok := colIDSet[col.ID]; ok {
-			return nil, fmt.Errorf("multiple assignments to same column \"%s\"", n.Column())
+			return nil, fmt.Errorf("multiple assignments to same column \"%s\"", n.Base)
 		}
 		colIDSet[col.ID] = struct{}{}
 		cols[i] = col

--- a/sql/testdata/insert
+++ b/sql/testdata/insert
@@ -443,3 +443,15 @@ CREATE TABLE nocols()
 
 statement error INSERT error: table nocols has \d+ columns but \d+ values were supplied
 INSERT INTO nocols VALUES (true, default)
+
+statement error column "kv" does not exist
+INSERT INTO kv (kv.k) VALUES ('hello')
+
+statement error column name "k\.\*" cannot be assigned to
+INSERT INTO kv (k.*) VALUES ('hello')
+
+statement error column name "k\[0\]" cannot be assigned to
+INSERT INTO kv (k[0]) VALUES ('hello')
+
+statement error column name "k\.v" cannot be assigned to
+INSERT INTO kv (k.v) VALUES ('hello')

--- a/sql/testdata/update
+++ b/sql/testdata/update
@@ -32,6 +32,18 @@ SELECT * FROM kv
 statement error column "m" does not exist
 UPDATE kv SET m = 9 WHERE k IN (1, 3)
 
+statement error column "kv" does not exist
+UPDATE kv SET kv.k = 9
+
+statement error column name "k\.\*" cannot be assigned to
+UPDATE kv SET k.* = 9
+
+statement error column name "k\[0\]" cannot be assigned to
+UPDATE kv SET k[0] = 9
+
+statement error column name "k\.v" cannot be assigned to
+UPDATE kv SET k.v = 9
+
 statement ok
 CREATE TABLE kv2 (
   k CHAR PRIMARY KEY,

